streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -209,7 +212,9 @@ void registerStateStores(final List<StateStore> allStores, final InternalProcess
         processorContext.uninitialize();
         for (final StateStore store : allStores) {
             if (stores.containsKey(store.name())) {
-                maybeRegisterStoreWithChangelogReader(store.name());
+                if (!stateUpdaterEnabled) {
+                    maybeRegisterStoreWithChangelogReader(store.name());
```

Comment:
QQ: this method is called `registerStateStores` but why does it expect `store` to be already in `stores`? The only place I can find `stores.put` is called is in `registerStore` and in that method, `maybeRegisterStoreWithChangelogReader` is called immediately after `stores.put` is called. So I'm confused what's the real purpose of this method and if `maybeRegisterStoreWithChangelogReader` call here is redundant.

Reply:
That is indeed confusing! When a task is recycled, the changelogs of the task to recycle are unregistered from the changelog reader. Then a new task is created and the state manager of the task to recycle is moved to the new task. For example, when a standby task is recycled to become an active task, its changelogs are unregistered from the changelog reader (https://github.com/apache/kafka/blob/2b2039f0ba88e210dd09031291c050cfcda7ce4a/streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java#L579), a new active task is created, and the state manager of the standby task is transferred to the new active task (https://github.com/apache/kafka/blob/2b2039f0ba88e210dd09031291c050cfcda7ce4a/streams/src/main/java/org/apache/kafka/streams/processor/internals/ActiveTaskCreator.java#L218). After the recycling, the new active task is in state `CREATED` and owns a state manager that has all state stores registered, but has the changelogs unregistered in the changelog reader. When the new active task is initialized, Streams tries to register the state stores again. That is when it reaches this code and it will execute the `if`-branch. That is the state store is registered but the changelogs aren't registered with the changelog reader.

Reply:
Another case for this is when we handle EOS task corruption due to no checkpoint file detected: in that case we would remove the corrupted task's changelogs, and re-initialize them. In this case the stores would be initialized but changelog readers not registered as well.

In the future we can decouple the registration of state stores and the `register(final TopicPartition partition, final ProcessorStateManager stateManager)` as long as in the latter case, we are sure that the `stateManager`'s stores map are already populated, which should always be true when the task is already in stateUpdater.